bpf/standalone/txg.py

```diff
@@ -0,0 +1,378 @@
+#!/usr/bin/env python3
+#
+# Copyright 2019 Delphix. All rights reserved.
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+#
+# This script prints out information about every spa_sync() for the domain0
+# zpool. This allows us to find a point in the past where I/O performance
+# degraded and some high-level symptoms about why it's slower now than it was,
+# which may be enough to correlate the change to a configuration /
+# environmental change.
+#
+
+from bcc import BPF, PerfType, PerfSWConfig
+import argparse
+import os
+import drgn
+from datetime import datetime
+
+# define BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/bpf_common.h>
+#include <uapi/linux/bpf.h>
+#include <uapi/linux/bpf_perf_event.h>
+#include <sys/spa_impl.h>
+#include <linux/sched.h>
+"""
+parser = argparse.ArgumentParser(
+        description='Collect spa_sync statistics for each txg.',
+        usage='estat txg [options]')
+parser.add_argument('-c', '--coll', type=int, action='store',
+                    dest='collection_sec',
+                    help='The collection interval in seconds')
+parser.add_argument('-p', '--pool', type=str, action='store',
+                    dest='pool',
+                    help='The pool to monitor (default: domain0)')
+args = parser.parse_args()
+
+if (args.pool):
+    bpf_text += '#define POOL "' + str(args.pool) + '"'
+else:
+    bpf_text += '#define POOL "domain0"'
+
+bpf_text += """
+typedef struct {
+    spa_t *spa;
+    u64 start;
+    u64 stop;
+    u64 time_since_last_sync;
+    u64 pool_sync_enter;
+    u64 elapsed_pass1;
+    u64 delay_entry_time;
+    u64 txg;
+    u64 throttle;
+    u64 delay_max;
+    u64 delay_sum;
+    u64 delay_count;
+} spa_sync_info;
+
+typedef struct {
+    u64  txg;
+    u64  time_since_last_sync;
+    u64  sync_time;
+    u64  elapsed_pass1;
+    u64  dirty_b;
+    u64  throttle;
+    u64  delay_max;
+    u64  delay_sum;
+    u64  delay_count;
+} spa_sync_data;
+
+typedef struct {
+    u64 *dirty_addr;
+    u64 dirty_max;
+} spa_sync_dirty;
+
+#define SPA_SYNC_DIRTY_INDEX  0;
+
+BPF_HASH(sync_info_map, u32, spa_sync_info);
+BPF_ARRAY(sync_dirty_map, spa_sync_dirty, 1);
+BPF_PERF_OUTPUT(sync_events);
+
+/*
+ * Store the spa_t address if the pool name matches POOL,
+ * "domain0" by default. Should only have to do the string
+ * comparision once.
+ */
+
+static inline void pool_set(spa_sync_info *info, spa_t *spa)
+{
+    char comparand[sizeof(POOL)];
+    bpf_probe_read(&comparand, sizeof(comparand), spa->spa_name);
+    char compare[] = POOL;
+    for (int i = 0; i < sizeof(comparand); ++i)
+       if (compare[i] != comparand[i])
```

Comment:
Does this check work if we have a pool named `domain0a`?

Reply:
Not as the default.  Would we like for all pools with names starting with "domain0" to be included?

Reply:
I verified what we talked about, that the loop includes the null terminator in the comparison.  So it only matches strings of the same length.

Reply:
I also tested a change to make a pool_compare function available for all the scripts in the BCCHelper.  It's a little more involved than you'd imagine as bcc has some restrictions on accessing maps in macros and include files.  So I think I'd rather include that in an upcoming PR.

Reply:
One approach for the SPA comparison would be to have the wrapper script run two underlying scripts. One that outputs the spa_t pointer that corresponds to the pool name (this could be an sdb or zdb command, for example), and then that pointer could get passed in as input to the bcc or bpftrace script. This reduces the complexity of having to factor this into the bcc or bpftrace script itself.

Reply:
Good idea.   
I'll update my arc_prefetch PR with another approach soon and we can carry this on there and decide which way to go.

Reply:
See [this PR](https://github.com/delphix/performance-diagnostics/pull/34)